theano/tensor/nnet/abstract_conv2d.py

```diff
@@ -0,0 +1,686 @@
+        # flops for any direction, sampling, padding, and border mode
+        inputs, filters = inp
+        outputs, = outp
+        assert inputs[1] == filters[1]
+        # nb mul and add by output pixel
+        flops = filters[2] * filters[3] * 2
+        # nb flops by output image
+        flops *= outputs[2] * outputs[3]
+        # nb patch multiplied
+        flops *= inputs[1] * filters[0] * inputs[0]
+        return flops
+
+
+
+class AbstractConv2d(BaseAbstractConv2d):
+    """
+    FIXME
+    """
+    def __init__(self,
+                 imshp=None,
+                 kshp=None,
+                 bsize=None,
+                 border_mode="valid",
+                 subsample=(1, 1),
+                 filter_flip = True):
+        super(AbstractConv2d, self).__init__(imshp, kshp, bsize,
+                                             border_mode, subsample, filter_flip)
+
+    def make_node(self, img, kern):
+        if img.type.ndim != 4:
+            raise TypeError('img must be 4D tensor')
+        if kern.type.ndim != 4:
+            raise TypeError('kern must be 4D tensor')
+
+        broadcastable=[img.broadcastable[0],
+                       kern.broadcastable[0],
+                       False, False]
+        #output = img.type.__class__(dtype=img.type.dtype,
+        #                            broadcastable=broadcastable)()
+        output = img.type.clone( broadcastable=broadcastable)()
+        return Apply(self, [img, kern], [output])
+
+    def perform(self, node, inp, out_):
+        raise NotImplementedError('AbstractConv2d theano optimization failed')
+
+    def grad(self, inp, grads):
+        bottom, weights = inp
+        top, = grads
+        d_bottom = AbstractConv2d_gradInputs(self.imshp, self.kshp,
+                                             self.bsize,
+                                             self.border_mode,
+                                             self.subsample,
+                                             self.filter_flip)(
+            weights, top, bottom.shape[-2:])
+        d_weights = AbstractConv2d_gradWeights(self.imshp, self.kshp,
+                                               self.bsize,
+                                               self.border_mode,
+                                               self.subsample,
+                                               self.filter_flip)(
+            bottom, top, weights.shape[-2:])
+        return d_bottom, d_weights
+
+
+class AbstractConv2d_gradWeights(BaseAbstractConv2d):
+    """Gradient wrt. filters for `AbstractConv2d`.
+
+    :note: You will not want to use this directly, but rely on
+           Theano's automatic differentiation or graph optimization to
+           use it as needed.
+
+    """
+
+    def __init__(self,
+                 imshp=None,
+                 kshp=None,
+                 bsize=None,
+                 border_mode="valid",
+                 subsample=(1, 1),
+                 filter_flip=True):
```

Comment:
In the grad Ops, we should document that border_mode, subsample and filter_flip are the flags that were used during the forward Op, not necessarily what has do be done during the execution of the conv-like code.